backend/service/topology/querybuilder.go

```diff
@@ -0,0 +1,181 @@
+package topology
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	sq "github.com/Masterminds/squirrel"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	topologyv1 "github.com/lyft/clutch/backend/api/topology/v1"
+)
+
+const (
+	column                   = "column"
+	columnIdentifer          = "column."
+	metadata                 = "metadata"
+	metadataIdentifer        = "metadata."
+	queryDefaultLimit uint64 = 100
+	maxResultLimit           = 1000
+)
+
+var metadataQueryValidator = regexp.MustCompile(`^[a-zA-Z0-9->'"]*$`)
+
+func paginatedQueryBuilder(
+	filter *topologyv1.SearchRequest_Filter,
+	sort *topologyv1.SearchRequest_Sort,
+	pageToken string,
+	limit uint64,
+) (sq.SelectBuilder, uint64, error) {
+	queryLimit := queryDefaultLimit
+	if limit > maxResultLimit {
+		return sq.SelectBuilder{}, 0, status.Error(codes.InvalidArgument, "maximum query limit is 1000")
+	} else if limit > 0 {
+		queryLimit = limit
+	}
+
+	// If no page is supplied default to 0
+	var pageNum uint64 = 0
+	var err error
+	if len(pageToken) > 0 {
+		pageNum, err = strconv.ParseUint(pageToken, 10, 64)
+		if err != nil {
+			return sq.SelectBuilder{}, 0, status.Error(codes.InvalidArgument, "unable to parse page_token")
+		}
+	}
+
+	queryOffset := pageNum * limit
+
+	query := sq.StatementBuilder.PlaceholderFormat(sq.Dollar).
+		Select("id", "data", "metadata").
+		From("topology_cache").
+		Limit(queryLimit).
+		Offset(queryOffset)
+
+	if filter != nil {
+		query, err = filterQueryBuilder(query, filter)
+		if err != nil {
+			return sq.SelectBuilder{}, 0, err
+		}
+	}
+
+	if sort != nil {
+		query, err = sortQueryBuilder(query, sort)
+		if err != nil {
+			return sq.SelectBuilder{}, 0, err
+		}
+	}
+
+	// Blindly increment pageNum by 1 for next_page_token
+	return query, pageNum + 1, nil
+}
+
+func filterQueryBuilder(query sq.SelectBuilder, f *topologyv1.SearchRequest_Filter) (sq.SelectBuilder, error) {
+	if f.Search != nil && len(f.Search.Field) > 0 {
+		searchIdentiferExpr := sq.Expr("id")
+		identifer, err := getFilterSortPrefixIdentifer(f.Search.Field)
+		if err != nil {
+			return sq.SelectBuilder{}, err
+		}
+
+		if identifer == column {
+			searchIdentiferExpr = sq.Expr(strings.TrimPrefix(f.Search.Field, columnIdentifer))
```

Comment:
this path is still injectable
```go
input: &topologyv1.SearchRequest_Filter{
    Search: &topologyv1.SearchRequest_Filter_Search{
        Field: "column.id NOT NULL;--",
        Text:  "cat",
    },
},

> SELECT * FROM topology_cache WHERE id NOT NULL;-- LIKE $1
```

echoing my above comment i would do overall validation on search field rather than nested in convertMetadataToQuery on the overall query.